src/core/contiguous_map.h

```diff
@@ -0,0 +1,109 @@
+// Copyright (c) 2014-2015 Dropbox, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef PYSTON_CORE_CONTIGUOUSMAP_H
+#define PYSTON_CORE_CONTIGUOUSMAP_H
+
+#include <llvm/ADT/DenseMap.h>
+#include <utility>
+#include <vector>
+
+namespace pyston {
+
+template <class TKey, class TVal, class TMap = llvm::DenseMap<TKey, int>> class ContiguousMap {
+    typedef TMap map_type;
+    typedef std::vector<TVal> vec_type;
+
+    map_type map;
+    vec_type vec;
+
+    std::vector<int> free_list;
+
+public:
+    typedef typename map_type::iterator iterator;
+    typedef typename map_type::const_iterator const_iterator;
+    typedef typename std::vector<TVal>::size_type size_type;
+
+    iterator find(TKey key) { return map.find(key); }
+
+    iterator begin() noexcept { return map.begin(); }
+    iterator end() noexcept { return map.end(); }
+
+    const_iterator begin() const noexcept { return map.begin(); }
+    const_iterator end() const noexcept { return map.end(); }
+
+    iterator erase(const_iterator position) {
+        int idx = map[position->first];
+        free_list.push_back(idx);
+        vec[idx] = TVal();
+        map.erase(position->first);
+        return begin(); // this is broken...
```

Comment:
Could this return the return value of map.erase?  Also, are we guaranteed that the provided iterator is in the map?